text/0000-safety-tags.md

```diff
@@ -0,0 +1,481 @@
+* Referencing a tag that has never been declared is also a **hard error**.
+
+[uninhabited]: https://doc.rust-lang.org/reference/glossary.html#uninhabited
+[type namespace]: https://doc.rust-lang.org/reference/names/namespaces.html
+[scopes]: https://doc.rust-lang.org/reference/names/scopes.html#item-scopes
+[visibility]: https://doc.rust-lang.org/reference/visibility-and-privacy.html
+[`UseTree`]: https://doc.rust-lang.org/reference/items/use-declarations.html
+
+Tags are treated as items so rustdoc can render their documentation and hyperlink tag references.
+And Rust-Analyzer can offer **full IDE support**: completion, go-to-definition/declaration, and
+doc-hover.
+
+Tags constitute a public API; therefore, any alteration to their declaration or definition must be
+evaluated against [Semantic Versioning][semver].
+* Adding a tag is a **minor** change.
+* Removing a tag is a **major** change.
+
+To give dependent crates time to migrate, mark obsolete tag items with `#[deprecated]`. Clippy will
+surface the deprecation warning whenever the tag is used.
+
+[semver]: https://doc.rust-lang.org/cargo/reference/semver.html
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Unstable Features
+
+Currently, safety tags requires the following unstable features
+* `#![feature(proc_macro_hygiene, stmt_expr_attributes)]` for tagging statements or expressions.
+* `#![feature(custom_inner_attributes)]` for `#![clippy::safety {}]` imports
+
+Since the safety-tag mechanism is implemented primarily in Clippy and Rust-Analyzer, no additional
+support is required from rustc.
+
+But we ask the libs team to adopt safety tags for all public `unsafe` APIs in libstd, along with
+their call sites. To enable experimentation, a nightly-only library feature
+`#![feature(safety_tags)]` should be introduced and remain unstable until the design is finalized.
+
+## Implementation in Clippy
+
+Procedure:
+
+1. Scan the crate for every item marked `#[clippy::safety::tag]`; cache the compiled tag metadata of
+   upstream dependencies under `target/` for later queries.
+2. Validate every `#![clippy::safety { ... }]` import by a reachability analysis that ensures every
+   referenced tag is defined and accessible.
+3. Verify that every unsafe call carries the required safety tags:
+   * Resolve the callee, collect its declared tags, then walk outward from the call site until the
+     function’s own signature confirms these tags are listed in its `#![clippy::safety { ... }]`
+     attribute.
+   * Tags are only discharged inside or onto an `unsafe fn`; it's an error to tag a safe function.
+   * If an unsafe call lacks any required tag, emit a diagnostic whose severity (warning or error)
+     is governed by the configured Clippy lint level.
+
+Libraries in `rust-lang/rust` must enforce tag checking as a hard error, guaranteeing that every tag
+definition and discharge is strictly valid.
+
+## Implementation in Rust Analyzer
```

Comment:
It feels weird to put the attributes under the `clippy` namespace if they are not a clippy-only thing. I can also expect other tools developed to analyze Rust will want to consider these attributes. I'd say put them under `diagnostics` or their own namespace.

Reply:
I've replaced clippy namespace with safety namespace. Restriction of single-path tool attr is solved by separating safety `requires` and `checked`. Thanks to @kennytm for the [idea](https://github.com/rust-lang/rfcs/pull/3842#discussion_r2247361847).